### ISSUE TITLE
characterbody3d doesnt interact with rigidbodies

### ISSUE THREAD

**Issue by djmaesen** [godotengine/godot#59473]
### Godot version

4.0 alpha5

### System information

linux

### Issue description

characterbody3d doesnt interact with rigid bodies
it treats them like kinematic bodies

### Steps to reproduce

try walking with a characterbody3d up to a dynamic rigidbody and it wont move

**Comment by Calinou** [comment]
@djmaesen Please upload a [minimal reproduction project](https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#provide-a-simple-example-project) to make this easier to troubleshoot.

**Comment by yosoyfreeman** [comment]
This seems like the expected result. Kinematic bodies are not supposed to influence any physic driven object unless you deliberately implement it. This predictability is the key element in kinematic bodies. 

Godot 4.0 include the possibility to automatically apply the velocity of a kinematic or physic body on the character as is a really common use case for moving platforms and kinda easy to implement. 

On the other hand, a lot of things can happen inside a kinematic body depending on your implementation, so is better to make this kind of things in the parts on the code you want to.

**Comment by djmaesen** [comment]
it worked in godot 3
my player interacted with rigidbodies. when calling move_and_slide

**Comment by yosoyfreeman** [comment]
> it worked in godot 3 my player interacted with rigidbodies. when calling move_and_slide

Ok, i understand you know.  you are talking about the infinite inertia property, which is not in godot 4.0.

**Comment by Zireael07** [comment]
IIRC you can switch it on, it's just not on by default unlike 3.x

**Comment by djmaesen** [comment]
> IIRC you can switch it on, it's just not on by default unlike 3.x

ok , how?

**Comment by yosoyfreeman** [comment]
> IIRC you can switch it on, it's just not on by default unlike 3.x

I may be wrong, but i remember talking about that feature being removed and i cant find any reference in the new docs:
[new documentation](https://docs.godotengine.org/en/latest/classes/class_characterbody3d.htmll) as i can find it on the old [kinematic body documentation](https://docs.godotengine.org/es/stable/classes/class_kinematicbody.html)

So, maybe in this case @djmaesen would like to make a new issue in [Godot proposals](https://github.com/godotengine/godot-proposals/issuesl) to discuss about the implementation of an equivalent feature if you like it. But is not technically a bug, but a missing feature.

**Comment by Zireael07** [comment]
Ooh, you may be right. Haven't used 4.x that much. I know most of the old flags were moved out of the move_and_collide() and move_and_slide() calls to elsewhere and thought it was the case with this too.

**Comment by yosoyfreeman** [comment]
> Ooh, you may be right. Haven't used 4.x that much. I know most of the old flags were moved out of the move_and_collide() and move_and_slide() calls to elsewhere and thought it was the case with this too.

I'm really deep in testing movement and sliding right now so i didn't get to the point of working on physics interactions, but thinking about it now it sounds like it could be a recurrent cause of doubts. While is not technically a bug, due to the amount of changes and the implementation of the feature to apply velocity for moving platforms, it would be great to have a section in the docs explaining that infinite_inertia has been removed, and ideally (But not so urgent) a little explanation about how to implement this things without mess with the built in feature for platforms.

I don't know ho is currently in charge of the documentation, but it would be great to, at least, let that person now about this.

**Comment by djmaesen** [comment]
> > IIRC you can switch it on, it's just not on by default unlike 3.x
> 
> I may be wrong, but i remember talking about that feature being removed and i cant find any reference in the [new documentation](https://docs.godotengine.org/en/latest/classes/class_characterbody3d.htmll) as i can find it on the old [kinematic body documentation](https://docs.godotengine.org/es/stable/classes/class_kinematicbody.html)
> 
> So, maybe in this case @djmaesen would like to make a new issue in [Godot proposals](https://github.com/godotengine/godot-proposals/issuesl) to discuss about the implementation of an equivalent feature if you like it. But is not technically a bug, but a missing feature.

so its not a bug, but a design decision? i think this is a bad idea. adding a function to the move_and_ slide() code would really be helpfull.

**Comment by Calinou** [comment]
Infinite inertia was removed from `master` by https://github.com/godotengine/godot/pull/51490.

**Comment by Zireael07** [comment]
Uuugh. Yet another case where it's eminently clear Godot 4 needs better docs, and fast, before more people pick it up and run into problems such as this.

**Comment by djmaesen** [comment]
so how do u apply "infinite inertia " in godot4 ? i dont understand the one directional collision layers

**Comment by Zireael07** [comment]
I don't either - it's one of the features that urgently need better docs.

**Comment by djmaesen** [comment]
please revert back to infinte inertia for 3d kinematic bodies. or at least add an option to enable it
i dont understand the tuse of directional collisionlayers.

**Comment by jk7c** [comment]
> please revert back to infinte inertia for 3d kinematic bodies. or at least add an option to enable it i dont understand the use of directional collisionlayers.

Here is a temporary solution:

```
func _push_rigidbody() -> void:
	if !enable_push_rigidbody: return

	var slideCollision
	for i in get_slide_collision_count():
		slideCollision = get_slide_collision(i)

	if slideCollision == null:
		return

	for i in slideCollision.get_collision_count():
		var collider = slideCollision.get_collider(i)
		if collider is RigidBody3D:
			collider.apply_central_impulse(-slideCollision.get_normal(i) * push_force)
	
```
In my case I get better results using a ShapeCast3D(CapsuleCast):

```
func _push_rigidbody() -> void:
	if !enable_push_rigidbody: return

	if _wall_helper_capsule != null:
		for i in _wall_helper_capsule.get_collision_count():
			var collider = _wall_helper_capsule.get_collider(i)
			if collider is RigidBody3D:
				collider.apply_central_impulse(-_wall_helper_capsule.get_collision_normal(i) * push_force)
```

**Comment by djmaesen** [comment]
i've already managed to get it working, it was as simple as setting up a correct collision layer interaction.

**Comment by LennDG** [comment]
> i've already managed to get it working, it was as simple as setting up a correct collision layer interaction.

Could you explain the collision layer interaction? I have a RigidBody3D with Collision mask and layer set to 1 and a CharacterBody3D set to the same but they are not interacting.

Obviously there are no docs on collision layers for Godot 4 (this is honestly a serious problem, I am going to start contributing there because it will be the death of the engine like this).

**Comment by CoreTaxxe** [comment]
> i've already managed to get it working, it was as simple as setting up a correct collision layer interaction.

How precisely do you set this up?
Have the player collision mask set to FALSE for rigidbodies you want to move (but the rigidbody mask for the player set to true)

Cause then the issue arises that the player doesn't interact with the rigidbody at all if the rigidbody is frozen.

**Comment by timshannon** [comment]
FWIW, this is the only way I was able to get this working, and it seems to work pretty well:

```c#
  public void ProcessCollisions()
    {
        for (var i = 0; i < GetSlideCollisionCount(); i++)
        {
            var collision = GetSlideCollision(i);
            for (var k = 0; k < collision.GetCollisionCount(); k++)
            {
                var body = collision.GetCollider(k) as RigidBody3D;
                if (body == null)
                    continue;
                var point = collision.GetPosition(k) - body.GlobalPosition;

                var force = 5; // put whatever force amount you want here

                body.ApplyImpulse(-collision.GetNormal(k) * force, point);
            }
        }
    }
```

**Comment by evan-gordon** [comment]
> > > i've already managed to get it working, it was as simple as setting up a correct collision layer interaction.
> > 
> > 
> > How precisely do you set this up? Have the player collision mask set to FALSE for rigidbodies you want to move (but the rigidbody mask for the player set to true)
> > Cause then the issue arises that the player doesn't interact with the rigidbody at all if the rigidbody is frozen.
> 
> it was relatively simple set world collision layer to 1, player to layer2 with mask to layer1(world) and then set the rigidbodies collision to layer 4 and mask to 1,2 and 4

I just spent 3 hours trying to figure this out after upgrading. Thanks for the clear solution!

That said, At minimum this needs better documentation. Ideally I'd like to see a more straightforward workaround before the launch of 4.0

**Comment by bend-n** [comment]
> it was relatively simple set world collision layer to 1, player to layer2 with mask to layer1(world) and then set the rigidbodies collision to layer 4 and mask to 1,2 and 4

I cant manage to stand on the rigid body though, it just slides oout.

**Comment by djmaesen** [comment]
> > it was relatively simple set world collision layer to 1, player to layer2 with mask to layer1(world) and then set the rigidbodies collision to layer 4 and mask to 1,2 and 4
> 
> I cant manage to stand on the rigid body though, it just slides oout.

me neither

**Comment by bend-n** [comment]
Sooo I guess the characterbody just has to manually push the rigid body.

**Comment by andersmmg** [comment]
Yes this is how it's always been, you have to choose one or the other or some people implement these interactions themselves

**Comment by iamjoeysox** [comment]
> 

Haha thank you a lot for this, saved me a bunch of headaches trying to figure out why the heck this worked fine in my project on Godot 3.x and suddenly broke. Not sure why others are posting needed hacks?

**Comment by bend-n** [comment]
@iamjoeysox 
> RigidBody2D sliding under a CharacterBody2D

Thats the entire reason the "hacks" are needed.
The hacks are actually the new intended method.

**Comment by iamjoeysox** [comment]
> @iamjoeysox
> 
> > RigidBody2D sliding under a CharacterBody2D
> 
> Thats the entire reason the "hacks" are needed. The hacks are actually the new intended method.

Haha yeah, I got myself super confused and figured out today I had it wrong lol. Thanks again!
